### DISCUSSION TITLE
Support for measuring libyears

### DISCUSSION THREAD

**Discussion by ZackKanter** [renovatebot/renovate#17387] [Suggest an Idea]
Given Renovate's focus on helping keep dependencies up-to-date, it would be fantastic if Renovate had the ability to calculate a repository's [libyear](https://libyear.com) statistics.

From libyear:

> A simple measure of software dependency freshness. It is a single number telling you how up-to-date your dependencies are.
> **Example**: Rails 5.0.0 (June 30, 2016) is roughly 1 libyear behind 5.1.2 (June 26, 2017).

One idea would be to have this calculated on the Dependency Dashboard.

There are packages available for many languages (though, ironically, they are mostly quite stale):

- dotnet (NuGet) - [dotnet-libyear](https://github.com/ecoAPM/dotnet-libyear)
- js - [libyear](https://github.com/jdanil/libyear) (GitHub Action: [libyear-node-action](https://github.com/s0/libyear-node-action))
- php (composer) - [php-libyear](https://github.com/ecoAPM/php-libyear)
- python (PyPI) - [libyear](https://github.com/nasirhjafri/libyear)
- python - [piprot ](https://github.com/sesh/piprot)(since 2013)
- ruby (bundler) - [libyear-bundler](https://github.com/jaredbeck/libyear-bundler)
- java (Gradle) - [libyear-gradle-plugin](https://github.com/f4lco/libyear-gradle-plugin)
- scala (Maven and SBT) - [release](https://github.com/novomind-ishop/release#show-libyears-of-dependency-updates)

**Comment by rarkins** [comment]
I'm interested in this idea, or related ones. I don't think libyear perfectly captures things though, for example let's say you are perfectly up to date with dependencies and then `lodash` releases `4.17.22`. Instantly you're 2 libyears behind the latest!

I think we'd want to implement the algorithm independently in TS within this repo so that we don't need to add the GPL-licensed JS library you linked to.

**Comment by ZackKanter** [comment]
> I'm interested in this idea, or related ones. 

Fantastic!

> I don't think libyear perfectly captures things though, for example let's say you are perfectly up to date with dependencies and then lodash releases 4.17.22. Instantly you're 2 libyears behind the latest!

I agree that libyear isn't perfect, but I actually think the case you described is a feature, not a bug – it ensures that signal isn't lost. Every dependency matters. While a more advanced algorithm could be devised, the value of libyear is in its simplicity and broad applicability.

> I think we'd want to implement the algorithm independently in TS within this repo so that we don't need to add the GPL-licensed JS library you linked to.

Makes sense.

**Comment by ZackKanter** [comment]
Two other concepts came to mind:

- https://github.com/marketplace/deadpendency
- https://github.com/ossf/scorecard

**Comment by MPV** [comment]
Found this issue today, and then quickly discovered that some kind of implementation has very recently started, i.e:
- https://github.com/renovatebot/renovate/pull/33521
